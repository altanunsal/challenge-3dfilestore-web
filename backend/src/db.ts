import { open } from "sqlite";
import { Database } from "sqlite3";

const db = open({ filename: ":memory:", driver: Database });

export async function runSQL<TReturn = void>(
  ...args: Parameters<Database["all"]>
): Promise<TReturn extends void ? void : TReturn[]> {
  try {
    const dbInstance = await db;
    return await dbInstance.all<TReturn extends void ? void : TReturn[]>(
      ...args
    );
  } catch (e) {
    console.error("DB ERROR:", e);
    throw e;
  }
}

(async () => {
  // Files table
  await runSQL(`
CREATE TABLE files (
  id VARCHAR(36) PRIMARY KEY,
  name VARCHAR(255),
  creation_date VARCHAR(255),
  size INTEGER,
  file_path VARCHAR(1023)
);
`);
})();
