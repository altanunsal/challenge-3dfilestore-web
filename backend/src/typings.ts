export type ObjFile = {
  id: string;
  name: string;
  creation_date: Date;
  size: number;
};

export type ObjFileMeta = Omit<ObjFile, "creation_date"> & {
  creation_date: string; // using string as the sqlite DB does expects ISO18601 date string
  file_path: string; // so we can retrieve the file from local
};

export type Vector3 = {
  x: number;
  y: number;
  z: number;
};
