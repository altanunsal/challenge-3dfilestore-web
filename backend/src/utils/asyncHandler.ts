import express from "express";

export const asyncHandler =
  (handler: express.RequestHandler): express.RequestHandler =>
  (req, res, next) => {
    Promise.resolve(handler(req, res, next)).catch(next);
  };
