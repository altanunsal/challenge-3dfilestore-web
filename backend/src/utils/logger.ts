import express from "express";

export const logger: express.RequestHandler = (req, res, next) => {
  res.on("finish", () => {
    console.info(
      `[${req.method.padStart(7, " ")}] ${req.path} - ${res.statusCode}`
    );
  });
  next();
};
