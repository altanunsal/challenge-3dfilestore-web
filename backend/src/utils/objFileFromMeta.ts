import { ObjFile, ObjFileMeta } from "../typings";

export function objFileFromMeta(meta: ObjFileMeta): ObjFile {
  const parsedDate = new Date(meta.creation_date);

  return {
    ...meta,
    creation_date: parsedDate,
  };
}
