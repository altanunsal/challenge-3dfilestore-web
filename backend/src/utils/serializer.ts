import { Vector3 } from "../typings";

export type SerializedVector = `${string} ${number} ${number} ${number}`;
export type DeserializedVector = {
  label: string;
  vector: Vector3;
};

export function isSerializedVector(
  queryParam: string
): queryParam is SerializedVector {
  const split = queryParam.split("  ");

  if (split.length !== 4) return false;

  let hasValidCoordinates = true;
  // split[0] is a label string
  [split[1], split[2], split[3]].forEach((point) => {
    if (isNaN(parseInt(point)) || isNaN(parseFloat(point))) {
      hasValidCoordinates = false;
    }
  });

  return hasValidCoordinates;
}

const SPLITTER: string = "  " as const;

function serialize(data: DeserializedVector): string {
  const {
    label,
    vector: { x, y, z },
  } = data;

  return [label, x, y, z].join(SPLITTER);
}

function deserialize(data: string): DeserializedVector {
  if (!isSerializedVector(data)) throw new Error("Unexpected format");

  const [label, x, y, z] = data.split(SPLITTER);
  return {
    label,
    vector: {
      x: Number(x),
      y: Number(y),
      z: Number(z),
    },
  };
}

export const serializer = {
  serialize,
  deserialize,
};
