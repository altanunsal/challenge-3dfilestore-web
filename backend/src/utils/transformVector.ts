import { Vector3 } from '../typings';

export function transformVector(vector: Vector3, scale: Vector3, offset: Vector3): Vector3 {
  const transformedVector: Vector3 = vector;

  transformedVector.x = vector.x * scale.x + offset.x;
  transformedVector.y = vector.y * scale.y + offset.y;
  transformedVector.z = vector.z * scale.z + offset.z;

  return transformedVector;
}
