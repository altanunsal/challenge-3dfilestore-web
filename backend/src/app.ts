import express from "express";
import cors from "cors";

import { downloadsRouter } from "./routes/downloads";
import { filesRouter } from "./routes/files";
import { transformRouter } from "./routes/transform";
import { logger } from "./utils/logger";

const app = express();

app.use(express.json());
app.use(cors());
app.use(logger);

const apiRoutes = express.Router();
apiRoutes.use("/files", filesRouter);
apiRoutes.use("/transform", transformRouter);
app.use("/api", apiRoutes);

app.use("/downloads", downloadsRouter);

app.use(function serverErrorHandler(
  err: unknown,
  req: express.Request,
  res: express.Response,
  next: express.NextFunction
) {
  console.error(err);
  res.status(500).send({
    error: err instanceof Error ? err.message || "Unknown error" : err,
  });
});

app.listen(3333);
