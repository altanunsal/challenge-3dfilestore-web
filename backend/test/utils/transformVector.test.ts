import { Vector3 } from '../../src/typings';
import { transformVector } from '../../src/utils/transformVector';

describe('transformVector', () => {
  it('transforms a vector according to the provided scale and offset', () => {
    const vector: Vector3 = {
      x: 2,
      y: 10,
      z: 6,
    };

    const scale: Vector3 = {
      x: 2,
      y: 10,
      z: 6,
    };

    const offset: Vector3 = {
      x: 1,
      y: 7,
      z: 8.4,
    };

    const result = transformVector(vector, scale, offset);

    expect(result).toMatchSnapshot();
  });
});
