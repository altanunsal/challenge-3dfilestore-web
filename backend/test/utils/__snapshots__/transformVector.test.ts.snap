// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformVector transforms a vector according to the provided scale and offset 1`] = `
{
  "x": 5,
  "y": 107,
  "z": 44.4,
}
`;
